axon-server-connector/src/main/java/org/axonframework/axonserver/connector/query/AxonServerQueryBus.java

```diff
@@ -246,11 +252,39 @@ public <R> Registration subscribe(@Nonnull String queryName,
                                            .queryChannel()
                                            .registerQueryHandler(localSegmentAdapter, queryDefinition);
 
-        return new AxonServerRegistration(localRegistration, serverRegistration::cancel);
+        queryHandlerNames.add(queryName);
+        return new AxonServerRegistration(() -> unsubscribe(queryName, localRegistration), serverRegistration::cancel);
```

Comment:
Thanks for this change, @rsobies! I think this flow will work out as discussed. 

And if somebody does enter the edge cases we discussed (so, somebody is directly interacting with the `localSegment` to subscribe handlers), we will find a way to deal with it at that point in time :+1: